Possible fix ajchellew/zwiftplay#7

ollypsilon opened the pull request:
Hi,
these changes might fix ajchellew/zwiftplay#7

![image](https://github.com/user-attachments/assets/47ce5d50-7335-4e57-8493-c4f194e5a73e)

Still some work to do, as I encounter unknown messages:
```text
Decrypt failed: Value cannot be null. (Parameter 'collection')
Unprocessed - Type: 42 Data: 2A 08 52 12 10 12 0E 08 12 10 00 18 90 03 22 05 FF 07 F0 FF 1F
```
But before I continue let's fix this state.

Best

Oliver

ollypsilon committed:
Update .gitignore

ollypsilon committed:
Update ZwiftPlayBleManager.cs

Add the IDisposable interface and Dispose method to the ZwiftPlayBleManager class.

RegisterCharacteristics handles TaskCanceledException that can occur during BLE characteristic registration and provides clear feedback about the connection state.

ConnectAsync method with proper null checks: checking for null references before proceeding with the BLE connection steps.
ConnectAsync method more robust by using async/await properly and adding error handling.

ollypsilon committed:
Update Program.cs

The RequestLEScanAsync method needs BluetoothLEScanOptions.

Handle the ObjectDisposedException by updating the scanning task.
Enhance the error handling in Program.cs by adding TaskCanceledException to catch blocks.

Add a null check for the ManufacturerData in the AdvertisementReceived event handler: This will ensure the code handles BLE advertisements without manufacturer data correctly and continues scanning for valid Zwift Play controllers.

ollypsilon committed:
Update .gitignore

dylix commented:
Awesome, this is pretty much what I had fixed, but I'm a terrible github user.. so nothing was ever submitted..

 I've been using more or less the same code for two weeks. I haven't had any issues with the 'unprocessed type 42' stuff.. and I know nothing of BT, so its blackmagic to me..

I've been using this console for indievelo to change virtual gears, change spotify music, brakes, u turn, etc.. fun times :) Glad my controllers have a use again!

ollypsilon committed:
Update ZwiftPlayConsoleApp.csproj

AllowUnsafeBlocks

ollypsilon committed:
Update BatteryStatus.cs

Unreachable part removed

ollypsilon committed:
Update AbstractZapDevice.cs

Minor changes

ollypsilon committed:
Added logging facility, some changes

Added logging class and methods

Moved some config parameters to new class

ollypsilon commented:
Just some changes to add a logging facility, helping to identify decrypt issue.

ollypsilon committed:
Moved SendKeys to configuration and command line

SendKeys no longer defined in code, instead moved to config and can be set on command line as well (--sendkeys)

ollypsilon commented:
Moved SendKeys to configuration and command line
SendKeys no longer defined in code, instead moved to config and can be set on command line as well (--sendkeys)

ollypsilon committed:
Indicate if left or right shoulder button

ollypsilon committed:
Added keyboard mapping to configuration

ollypsilon commented:
Added keyboard mapping to configuration

ollypsilon committed:
Config moved to JSON file, mapping file added

- Moved configuration to JSON file.
-- Moved BleScanConfig part into AppSettings.
- Added MappingFile optionto use a JSON file which contains keyboard mapping.
-- Added default TPVirtual.json file.
-- Keyboard mapping allows shift key combinations to use
- More robust BT device implementation with cleanup
-- Possible ToDo: implement BLE retry logic
- App class now IDisposable

ollypsilon commented:
- Moved configuration to JSON file.
  - Moved BleScanConfig part into AppSettings.
- Added MappingFile optionto use a JSON file which contains keyboard mapping.
  - Added default TPVirtual.json file.
  - Keyboard mapping allows shift key combinations to use
- More robust BT device implementation with cleanup
  - Possible ToDo: implement BLE retry logic
- App class now IDisposable

ollypsilon committed:
Target framework now net8.0-windows10.0.19041.0 for continued security updates and support

ollypsilon committed:
Added logging for KeyboardKeys class, clean up

pauloserra81 commented:
Just found this, and it is exactly the code I wanted !
Just can't get it to work, even with the updated version is still does nothing here.

Controllers connect, and then nothing.

Any hints to debug it ? 

I am using play controllers 1.3.1 if it matters ...

ollypsilon commented:
> Just found this, and it is exactly the code I wanted ! Just can't get it to work, even with the updated version is still does nothing here.
> 
> Controllers connect, and then nothing.
> 
> Any hints to debug it ?
> 
> I am using play controllers 1.3.1 if it matters ...

Hi!
Well, much to less information to be able to help you.
> Controllers connect [...]

So, controllers connect, meaning from blinking they change to lit?
If so, what happens if you e.g. open a file with an editor and you press some keys on controller?
Keys are mapped in this version, so cursor buttons act as cursor buttons on keyboard. Do they work in editor?

> [...] ,and then nothing.

What do you expect, what should happen?

> Any hints to debug it ?

Enable debug settings in config file.

pauloserra81 commented:
Nevermind. I got it solved.

Used a different branch named "bt debug" and everthing was working there.
Just did some updates to my use case, and got it working.

Thanks !

joppez commented:
How do I get the Windows App started? I downloaded the zip and did run the .bat. But I get some errors and I cannot find an .exe file to start. I appreciate any help.

ollypsilon commented:
> How do I get the Windows App started? I downloaded the zip and did run the .bat. But I get some errors and I cannot find an .exe file to start. I appreciate any help.

As the merge request hasn't been validated by the author yet, the Zip contains the old code still. You will have to download the code and compile it to get an exe to run.
